core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
         }
     }
 
+    private void maybeRefreshEngine() {
+        if (indexSettings.getRefreshInterval().millis() > 0) {
+            for (IndexShard shard : this.shards.values()) {
+                switch (shard.state()) {
+                    case CREATED:
+                    case RECOVERING:
+                    case CLOSED:
+                        continue;
+                    case POST_RECOVERY:
+                    case STARTED:
+                    case RELOCATED:
+                        try {
+                            shard.refresh("schedule");
+                        } catch (EngineClosedException | AlreadyClosedException ex) {
+                            // fine - continue;
+                        }
+                        continue;
+                    default:
+                        throw new IllegalStateException("unknown state: " + shard.state());
+                }
+            }
+        }
+    }
 
-    /**
-     * FSyncs the translog for all shards of this index in a defined interval.
-     */
-    final static class AsyncTranslogFSync implements Runnable {
-        private final IndexService indexService;
-        private final ThreadPool threadPool;
+    static abstract class BaseAsyncTask implements Runnable, Closeable {
```

Comment:
Should this subclass `AbstractRunnable` so we are graceful about handling rejections/failures from the thread pool?